core/controllers/reader.py

```diff
@@ -119,6 +119,78 @@ def classify_string_classifier_rule(state, normalized_answer):
     return None
 
 
+def _get_exploration_player_data(self, exploration_id, version, collection_id):
```

Comment:
Oh, I see -- instead of passing self, I suggest calculating `can_edit` in the individual handlers. Also, if you need to raise an exception, I suggest just `raise Exception`, catch that in the individual handlers, and transform it to PageNotFoundException. `self` is a lot of information to pass around, and it's better to keep this function tight and well-defined.